design/api_design.md

```diff
@@ -0,0 +1,64 @@
+# API design
+
+We will follow standard as possible for our REST APIs:
+
+- CRUD
+- Search API
+
+## Global
+
+For all APIs, a JSON will be returned:
+- in case of success, pertinent data depending of the API's purpose
+- in case of error, we will have the following parameters:
+    - `detail`: message that details the error
+
+## CRUD
+
+At least four APIs by database's table to Create, Read, Update and Delete data.
+
+## Search API
+
+At least one search API by database's table with at least pagination, filter and sorting.
+
+The pagination will be manage as following:
+
+- query parameters for pagination:
+    1. `page`: indicates the current page number
+        - starting at 1 by default
+        - e.g., `?page=1`
+    2. `limit`: specifies the number of items to be returned per page
+        - at 20 by default
+        - at 100 maximum
+        - e.g., `?limit=50`
+    3. `sort` (optional): allows sorting by one or more fields
+        - `+` for ascending, `-` for descending
+        - e.g., `?sort=+name`
+    4. `filter` (optional): applies filters to refine results
+        - e.g., `?filter[archived]=false`
+    5. date filtering: allows filtering by date ranges
+        - exact parameters will depends of the API research possibilities
+        - format expected:
+            - `start_<date_parameter>`: filter results from this date
+            - `end_<date_parameter>`: filters results up to and including this date
+        - e.g., `?start_date=2024-01-01&end_date=2024-12-31`
+        - we will follow ISO 8601 format:
+            - date format: YYYY-MM-DD
+            - datetime format: YYYY-MM-DDTHH:MM:SSZ
+- response structure:
+    1. `data`: contains the array of items for the current page
+    2. `meta`:
+        - `total_items`: total number of items available in the dataset
```

Comment:
Is it necessary to return query parameters ?

Reply:
total_items is not the query parameters, it's what you called count in https://github.com/denis-jullien/leafy-ils/issues/43

Reply:
The decision to include a meta section in the response structure instead of placing pagination parameters directly at the JSON root comes down to organization, clarity, and scalability. Here’s why having a meta section is often preferred.

Reply:
Encapsulating pagination parameters and other metadata in a meta section:

- Ensures clean and scalable response structures.
- Avoids naming conflicts or root-level clutter.
- Aligns with widely accepted standards like JSON:API.
- Simplifies future API evolution.